Automattic-Tracks-iOS/Crash Logging/CrashLogging.swift

```diff
@@ -5,188 +5,259 @@ import CocoaLumberjack
+    /// - Parameters:
+    ///   - error: The error object
+    ///   - userInfo: A dictionary containing additional data about this error.
+    ///   - level: The level of severity to report in Sentry (`.error` by default)
+    func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error) {
 
-     - Parameters:
-     - error: The error object
-     - userInfo: A dictionary containing additional data about this error.
-     - level: The level of severity to report in Sentry (`.error` by default)
-    */
-    static func logError(_ error: Error, userInfo: [String: Any]? = nil, level: SentrySeverity = .error) {
-        let event = Event(level: .error)
-        event.message = error.localizedDescription
+        let event = Event(level: level)
+
+        /// Use the unlocalized error message for better grouping
+        event.message = SentryMessage(formatted: (error as NSError).description)
+
+        /// If the developer provides their own userInfo, use that – otherwise read it from the Error
         event.extra = userInfo ?? (error as NSError).userInfo
+
+        /// Attach the localized description in case it has additional data
+        event.extra?["localized-description"] = error.localizedDescription
+
+        event.timestamp = Date()
+
+        SentrySDK.capture(event: event)
+        dataProvider.didLogErrorCallback?(event)
+    }
+
+    /// Writes a message to the Crash Logging system, and includes a stack trace.
+    ///
+    /// - Parameters:
+    ///   - message: The message
+    ///   - properties: A dictionary containing additional information about this error
+    ///   - level: The level of severity to report in Sentry (`.info` by default)
+    func logMessage(_ message: String, properties: [String: Any]? = nil, level: SentryLevel = .info) {
+
+        let event = Event(level: level)
+        event.message = SentryMessage(formatted: message)
+        event.extra = properties
         event.timestamp = Date()
 
-        Client.shared?.snapshotStacktrace {
-            Client.shared?.appendStacktrace(to: event)
+        SentrySDK.capture(event: event)
+        dataProvider.didLogMessageCallback?(event)
+    }
+
+    /// Sends an `Event` to Sentry and triggers a callback on completion
+    func logErrorImmediately(_ error: Error, userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+        try logErrorsImmediately([error], userInfo: userInfo, level: level, callback: callback)
+    }
+
+    func logErrorsImmediately(_ errors: [Error], userInfo: [String: Any]? = nil, level: SentryLevel = .error, callback: @escaping (Result<Bool, Error>) -> Void) throws {
+
+        var serializer = SentryEventSerializer(dsn: dataProvider.sentryDSN)
+
+        errors.forEach {
+            let event = Event(level: level)
+            event.message = SentryMessage(formatted: $0.localizedDescription)
```

Comment:
Some additional insight from Playground testing:

```swift
import Foundation

enum MyError: Error {
    case foo(code: Int)
}

let e = MyError.foo(code: 42)

print((e as NSError).description) // __lldb_expr_9.MyError.foo(code: 42)
print(String(describing: e)) // foo(code: 42)
print(e.localizedDescription) // The operation couldn’t be completed. (__lldb_expr_9.MyError error 0.)
```

Notice how the last one is what we end up seeing in Sentry's dashboard, because the `SentryTestError` we use in our example project doesn't have a custom `errorDescription: String?` implementation, despite being declared as `LocalizedError` (the default implementation for `LocalizedError.errorDescription` being what you see in the last line of my playground sample above)